File: go/internal/cmd/stats.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/spf13/cobra"
	"gopkg.in/yaml.v3"

	"git.zib.de/cspiegel/threads/internal/output"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	statsDownVal   int
	statsRecursive bool
	statsUpVal     int
	statsFormat    string
	statsJSON      bool
)

var statsCmd = &cobra.Command{
	Use:   "stats [path]",
	Short: "Show thread count by status",
	Long: `Show thread count by status at the specified level.

Path resolution:
  (none)  → PWD (current directory)
  .       → PWD (explicit)
  ./X/Y   → PWD-relative
  /X/Y    → Absolute
  X/Y     → Git-root-relative

Use -d/--down to include subdirectories, -u/--up to include parent directories.
Use -r as an alias for --down (unlimited depth).`,
	Args: cobra.MaximumNArgs(1),
	RunE: runStats,
}

func init() {
	statsCmd.Flags().IntVarP(&statsDownVal, "down", "d", -1, "Search subdirectories (N levels, 0=unlimited)")
	statsCmd.Flags().BoolVarP(&statsRecursive, "recursive", "r", false, "Alias for --down (unlimited depth)")
	statsCmd.Flags().IntVarP(&statsUpVal, "up", "u", -1, "Search parent directories (N levels, 0=to git root)")
	statsCmd.Flags().StringVarP(&statsFormat, "format", "f", "fancy", "Output format: fancy, plain, json, yaml")
	statsCmd.Flags().BoolVar(&statsJSON, "json", false, "Output as JSON (shorthand for --format=json)")
}

// statsSearchDirection describes the search direction for stats output display.
type statsSearchDirection struct {
	hasDown   bool
	downDepth int
	hasUp     bool
	upDepth   int
}

func (s *statsSearchDirection) description() string {
	var parts []string

	if s.hasDown {
		if s.downDepth < 0 {
			parts = append(parts, "recursive")
		} else {
			parts = append(parts, fmt.Sprintf("down %d", s.downDepth))
		}
	}

	if s.hasUp {
		if s.upDepth < 0 {
			parts = append(parts, "up")
		} else {
			parts = append(parts, fmt.Sprintf("up %d", s.upDepth))
		}
	}

	if len(parts) == 0 {
		return ""
	}
	return fmt.Sprintf(" (%s)", strings.Join(parts, ", "))
}

func (s *statsSearchDirection) isSearching() bool {
	return s.hasDown || s.hasUp
}

type statusCount struct {
	Status string `json:"status" yaml:"status"`
	Count  int    `json:"count" yaml:"count"`
}

type sortedCount struct {
	Key   string
	Value int
}

func runStats(cmd *cobra.Command, args []string) error {
	gitRoot := getWorkspace()

	// Determine output format (handle --json shorthand)
	var format output.Format
	if statsJSON {
		format = output.FormatJSON
	} else {
		format, _ = output.ParseFormat(statsFormat)
		format = format.Resolve()
	}

	// Parse path filter
	pathArg := ""
	if len(args) > 0 {
		pathArg = args[0]
	}

	// Resolve the scope
	scope, err := workspace.InferScope(gitRoot, pathArg)
	if err != nil {
		return err
	}
	filterPath := scope.Path
	startPath := scope.ThreadsDir[:len(scope.ThreadsDir)-len("/.threads")]

	// Determine search direction
	downSet := cmd.Flags().Changed("down")
	hasDown := downSet || statsRecursive
	downDepth := -1
	if downSet && statsDownVal > 0 {
		downDepth = statsDownVal
	}

	upSet := cmd.Flags().Changed("up")
	hasUp := upSet
	upDepth := -1
	if upSet && statsUpVal > 0 {
		upDepth = statsUpVal
	}

	// Build options
	options := workspace.NewFindOptions()

	if hasDown {
		depth := downDepth
		if depth < 0 {
			depth = 0
		}
		options = options.WithDown(&depth)
	}

	if hasUp {
		depth := upDepth
		if depth < 0 {
			depth = 0
		}
		options = options.WithUp(&depth)
	}

	// Track search direction for output
	searchDir := &statsSearchDirection{
		hasDown:   hasDown,
		downDepth: downDepth,
		hasUp:     hasUp,
		upDepth:   upDepth,
	}

	// Find threads using options
	threads, err := workspace.FindThreadsWithOptions(startPath, gitRoot, options)
	if err != nil {
		return err
	}

	counts := make(map[string]int)
	total := 0

	for _, path := range threads {
		relPath := workspace.ParseThreadPath(gitRoot, path)

		// Path filter: if not searching, only count threads at the specified level
		if !searchDir.isSearching() {
			if relPath != filterPath {
				continue
			}
		}

		t, err := thread.Parse(path)
		if err != nil {
			continue
		}

		status := t.BaseStatus()
		if status == "" {
			status = "(none)"
		}

		counts[status]++
		total++
	}

	// Sort by count descending
	var sorted []sortedCount
	for k, v := range counts {
		sorted = append(sorted, sortedCount{k, v})
	}
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Value > sorted[j].Value
	})

	switch format {
	case output.FormatFancy:
		return statsOutputFancy(sorted, total, filterPath, searchDir)
	case output.FormatPlain:
		return statsOutputPlain(sorted, total, gitRoot, filterPath, searchDir)
	case output.FormatJSON:
		return statsOutputJSON(sorted, total, gitRoot, filterPath)
	case output.FormatYAML:
		return statsOutputYAML(sorted, total, gitRoot, filterPath)
	default:
		return statsOutputFancy(sorted, total, filterPath, searchDir)
	}
}

func statsOutputFancy(sorted []sortedCount, total int, filterPath string, searchDir *statsSearchDirection) error {
	pathDesc := filterPath
	if filterPath == "." {
		pathDesc = "repo root"
	}

	searchSuffix := searchDir.description()

	fmt.Printf("Stats for threads in %s%s\n", pathDesc, searchSuffix)
	fmt.Println()

	if total == 0 {
		fmt.Println("No threads found.")
		if !searchDir.isSearching() {
			fmt.Println("Hint: use -r to include nested directories, -u to search parents")
		}
		return nil
	}

	fmt.Println("| Status     | Count |")
	fmt.Println("|------------|-------|")
	for _, kv := range sorted {
		fmt.Printf("| %-10s | %5d |\n", kv.Key, kv.Value)
	}
	fmt.Println("|------------|-------|")
	fmt.Printf("| %-10s | %5d |\n", "Total", total)

	return nil
}

func statsOutputPlain(sorted []sortedCount, total int, gitRoot, filterPath string, searchDir *statsSearchDirection) error {
	pwd, _ := os.Getwd()
	fmt.Printf("PWD: %s\n", pwd)
	fmt.Printf("Git root: %s\n", gitRoot)
	fmt.Println()

	pathDesc := filterPath
	if filterPath == "." {
		pathDesc = "repo root"
	}

	searchSuffix := searchDir.description()

	fmt.Printf("Stats for threads in %s%s\n", pathDesc, searchSuffix)
	fmt.Println()

	if total == 0 {
		fmt.Println("No threads found.")
		if !searchDir.isSearching() {
			fmt.Println("Hint: use -r to include nested directories, -u to search parents")
		}
		return nil
	}

	fmt.Println("| Status     | Count |")
	fmt.Println("|------------|-------|")
	for _, kv := range sorted {
		fmt.Printf("| %-10s | %5d |\n", kv.Key, kv.Value)
	}
	fmt.Println("|------------|-------|")
	fmt.Printf("| %-10s | %5d |\n", "Total", total)

	return nil
}

func statsOutputJSON(sorted []sortedCount, total int, gitRoot, filterPath string) error {
	type jsonOutput struct {
		GitRoot string        `json:"git_root"`
		Path    string        `json:"path"`
		Counts  []statusCount `json:"counts"`
		Total   int           `json:"total"`
	}

	var counts []statusCount
	for _, kv := range sorted {
		counts = append(counts, statusCount{
			Status: kv.Key,
			Count:  kv.Value,
		})
	}

	output := jsonOutput{
		GitRoot: gitRoot,
		Path:    filterPath,
		Counts:  counts,
		Total:   total,
	}

	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	return enc.Encode(output)
}

func statsOutputYAML(sorted []sortedCount, total int, gitRoot, filterPath string) error {
	type yamlOutput struct {
		GitRoot string        `yaml:"git_root"`
		Path    string        `yaml:"path"`
		Counts  []statusCount `yaml:"counts"`
		Total   int           `yaml:"total"`
	}

	var counts []statusCount
	for _, kv := range sorted {
		counts = append(counts, statusCount{
			Status: kv.Key,
			Count:  kv.Value,
		})
	}

	output := yamlOutput{
		GitRoot: gitRoot,
		Path:    filterPath,
		Counts:  counts,
		Total:   total,
	}

	enc := yaml.NewEncoder(os.Stdout)
	enc.SetIndent(2)
	return enc.Encode(output)
}

File: src/cmd/stats.rs
use std::collections::HashMap;
use std::path::Path;

use clap::Args;
use colored::Colorize;
use serde::Serialize;
use tabled::settings::Style;
use tabled::{Table, Tabled};

use crate::args::{DirectionArgs, FilterArgs, FormatArgs};
use crate::config::{Config, is_quiet, root_name};
use crate::output::{self, OutputFormat};
use crate::thread::{self, Thread};
use crate::workspace;

#[derive(Args)]
pub struct StatsArgs {
    /// Path to show stats for (git-root-relative, ./pwd-relative, or absolute)
    #[arg(default_value = "")]
    path: String,

    #[command(flatten)]
    direction: DirectionArgs,

    #[command(flatten)]
    filter: FilterArgs,

    #[command(flatten)]
    format: FormatArgs,
}

#[derive(Serialize)]
struct StatusCount {
    status: String,
    count: usize,
}

pub fn run(args: StatsArgs, git_root: &Path, config: &Config) -> Result<(), String> {
    let format = args.format.resolve();

    // Parse path filter
    let path_filter = if args.path.is_empty() {
        None
    } else {
        Some(args.path.as_str())
    };

    // Resolve the scope
    let scope = workspace::infer_scope(git_root, path_filter)?;
    let filter_path = scope.path.clone();
    let start_path = scope.threads_dir.parent().unwrap_or(git_root);

    // Convert direction args to find options
    let options = args.direction.to_find_options();

    // Find threads using options
    let threads = workspace::find_threads_with_options(start_path, git_root, &options)?;

    let mut counts: HashMap<String, usize> = HashMap::new();
    let mut total = 0;

    for path in threads {
        let rel_path = workspace::parse_thread_path(git_root, &path);

        // Path filter: if not searching, only count threads at the specified level
        if !args.direction.is_searching() && rel_path != filter_path {
            continue;
        }
        // Note: find_threads_with_options already handles direction/depth filtering

        let t = match Thread::parse(&path) {
            Ok(t) => t,
            Err(_) => continue,
        };

        let status = t.base_status();

        // Filter out closed threads unless --include-closed
        if !args.filter.include_closed() && thread::is_closed(&status) {
            continue;
        }

        let status = if status.is_empty() {
            "(none)".to_string()
        } else {
            status
        };

        *counts.entry(status).or_insert(0) += 1;
        total += 1;
    }

    // Sort by count descending
    let mut sorted: Vec<_> = counts.into_iter().collect();
    sorted.sort_by(|a, b| b.1.cmp(&a.1));

    let include_closed = args.filter.include_closed();

    match format {
        OutputFormat::Pretty => output_pretty(
            &sorted,
            total,
            &filter_path,
            &args.direction,
            include_closed,
            config,
        ),
        OutputFormat::Plain => output_plain(
            &sorted,
            total,
            git_root,
            &filter_path,
            &args.direction,
            include_closed,
            config,
        ),
        OutputFormat::Json => output_json(&sorted, total, git_root, &filter_path),
        OutputFormat::Yaml => output_yaml(&sorted, total, git_root, &filter_path),
    }
}

/// Row data for stats table
#[derive(Tabled)]
struct StatsRow {
    #[tabled(rename = "STATUS")]
    status: String,
    #[tabled(rename = "COUNT")]
    count: String,
}

/// Build filter description for summary line
fn build_filter_desc(include_closed: bool, direction: &DirectionArgs) -> String {
    let mut parts = Vec::new();

    if !include_closed {
        parts.push("open".to_string());
    } else {
        parts.push("all statuses".to_string());
    }

    let dir_desc = direction.description();
    if !dir_desc.is_empty() {
        parts.push(dir_desc);
    }

    parts.join(", ")
}

fn output_pretty(
    sorted: &[(String, usize)],
    total: usize,
    filter_path: &str,
    direction: &DirectionArgs,
    include_closed: bool,
    config: &Config,
) -> Result<(), String> {
    let path_desc = if filter_path == "." {
        root_name(config).to_string()
    } else {
        filter_path.to_string()
    };

    let filter_desc = build_filter_desc(include_closed, direction);

    println!(
        "{} {} ({})",
        "Stats for threads in".bold(),
        path_desc,
        filter_desc.dimmed()
    );
    println!();

    if total == 0 {
        println!("{}", "No threads found.".dimmed());
        if !direction.is_searching() && !is_quiet(config) {
            println!(
                "{}",
                "Hint: use --down to include nested directories, --up to search parents".dimmed()
            );
        }
        return Ok(());
    }

    // Build table rows with styled status
    let mut rows: Vec<StatsRow> = sorted
        .iter()
        .map(|(status, count)| StatsRow {
            status: output::style_status(status).to_string(),
            count: count.to_string(),
        })
        .collect();

    // Add total row
    rows.push(StatsRow {
        status: "Total".bold().to_string(),
        count: total.to_string().bold().to_string(),
    });

    let mut table = Table::new(rows);
    table.with(Style::rounded());

    println!("{}", table);

    Ok(())
}

fn output_plain(
    sorted: &[(String, usize)],
    total: usize,
    git_root: &Path,
    filter_path: &str,
    direction: &DirectionArgs,
    include_closed: bool,
    config: &Config,
) -> Result<(), String> {
    let pwd = std::env::current_dir()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| "(unknown)".to_string());

    println!("PWD: {}", pwd);
    println!("Git root: {}", git_root.display());
    println!();

    let path_desc = if filter_path == "." {
        root_name(config).to_string()
    } else {
        filter_path.to_string()
    };

    let filter_desc = build_filter_desc(include_closed, direction);

    println!("Stats for threads in {} ({})", path_desc, filter_desc);
    println!();

    if total == 0 {
        println!("No threads found.");
        if !direction.is_searching() && !is_quiet(config) {
            println!("Hint: use --down to include nested directories, --up to search parents");
        }
        return Ok(());
    }

    // Plain pipe-delimited format
    println!("STATUS | COUNT");
    for (status, count) in sorted {
        println!("{} | {}", status, count);
    }
    println!("Total | {}", total);

    Ok(())
}

fn output_json(
    sorted: &[(String, usize)],
    total: usize,
    git_root: &Path,
    filter_path: &str,
) -> Result<(), String> {
    #[derive(Serialize)]
    struct JsonOutput {
        git_root: String,
        path: String,
        counts: Vec<StatusCount>,
        total: usize,
    }

    let counts: Vec<StatusCount> = sorted
        .iter()
        .map(|(status, count)| StatusCount {
            status: status.clone(),
            count: *count,
        })
        .collect();

    let output = JsonOutput {
        git_root: git_root.to_string_lossy().to_string(),
        path: filter_path.to_string(),
        counts,
        total,
    };

    let json = serde_json::to_string_pretty(&output)
        .map_err(|e| format!("JSON serialization failed: {}", e))?;
    println!("{}", json);
    Ok(())
}

fn output_yaml(
    sorted: &[(String, usize)],
    total: usize,
    git_root: &Path,
    filter_path: &str,
) -> Result<(), String> {
    #[derive(Serialize)]
    struct YamlOutput {
        git_root: String,
        path: String,
        counts: Vec<StatusCount>,
        total: usize,
    }

    let counts: Vec<StatusCount> = sorted
        .iter()
        .map(|(status, count)| StatusCount {
            status: status.clone(),
            count: *count,
        })
        .collect();

    let output = YamlOutput {
        git_root: git_root.to_string_lossy().to_string(),
        path: filter_path.to_string(),
        counts,
        total,
    };

    let yaml =
        serde_yaml::to_string(&output).map_err(|e| format!("YAML serialization failed: {}", e))?;
    print!("{}", yaml);
    Ok(())
}
